core/common/src/main/java/alluxio/conf/path/PathConfiguration.java

```diff
@@ -0,0 +1,36 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.conf.path;
+
+import alluxio.AlluxioURI;
+import alluxio.conf.AlluxioConfiguration;
+import alluxio.conf.PropertyKey;
+
+import java.util.Optional;
+
+/**
+ * Path level configuration.
+ *
+ * An {@link AlluxioConfiguration} is set for each path pattern. This class matches an Alluxio
+ * path against the patterns, then from the configurations of the matched patterns, it chooses the
+ * configuration that contains the specified property key and belongs to the best matched pattern.
+ *
+ * Different implementations have different definitions of the best match.
+ */
+public interface PathConfiguration {
+  /**
+   * @param path the Alluxio path
+   * @param key the property key
+   * @return the chosen configuration matching the path and containing the key
+   */
+  Optional<AlluxioConfiguration> getConfiguration(AlluxioURI path, PropertyKey key);
```

Comment:
What's the reasoning for returning the entire configuration object? We need to call `getConfiguration` every time we need a different key, so we don't get to reuse the Configuration object.

Reply:
Ideally, the methods in this class should have be like `String get(AlluxioURI path, PropertyKey key)`, `int getInt(AlluxioURI path, PropertyKey key)`, `boolean getBool(xxxx)`, `Enum getEnum(xxx)`, basically, it should contain similar methods in `AlluxioConfiguration` with the `path` parameter.

But the current design makes this interface cleaner and serves the purpose well currently.